lib/std/start.zig

```diff
@@ -218,21 +220,28 @@ fn EfiMain(handle: uefi.Handle, system_table: *uefi.tables.SystemTable) callconv
         uefi.Status => {
             return @intFromEnum(root.main());
         },
-        uefi.Error!void => {
-            root.main() catch |err| switch (err) {
-                error.Unexpected => @panic("EfiMain: unexpected error"),
-                else => {
-                    const status = uefi.Status.fromError(@errorCast(err));
-                    return @intFromEnum(status);
-                },
+        else => |T| if (@typeInfo(T) == .error_union) {
+            const result = root.main() catch |err| {
+                std.log.err("{s}", .{@errorName(err)});
+                if (@errorReturnTrace()) |trace| {
+                    std.debug.dumpStackTrace(trace.*);
+                }
+                return @intFromEnum(uefi.Status.fromError(@errorCast(err)));
             };
 
-            return 0;
+            switch (@TypeOf(result)) {
+                void => return 0,
+                u8, usize => {
+                    return result;
+                },
+                uefi.Status => {
+                    return @intFromEnum(result);
+                },
+                else => @compileError(bad_efi_main_ret),
+            }
```

Comment:
We're not reintroducing untyped u8/usize return values. Keep this as is and only touch the existing `switch (err) {}`.